File: sjf(non).c
#include <stdio.h>
#include <stdbool.h>
#include <limits.h>

typedef struct {
    char pid[5];         // Process ID
    int arrival_time;    // Arrival Time
    int burst_time;      // Burst Time
    int completion_time; // Completion Time
    int turnaround_time; // Turnaround Time
    int waiting_time;    // Waiting Time
    int response_time;   // Response Time
    bool completed;      // Process Completion Status
} Process;

// Function to sort processes based on arrival time, and burst time for ties
void sort_by_arrival_and_burst(Process processes[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (processes[j].arrival_time > processes[j + 1].arrival_time ||
                (processes[j].arrival_time == processes[j + 1].arrival_time &&
                 processes[j].burst_time > processes[j + 1].burst_time)) {
                Process temp = processes[j];
                processes[j] = processes[j + 1];
                processes[j + 1] = temp;
            }
        }
    }
}

// Function to find the next process to execute
int find_next_process(Process processes[], int n, int current_time) {
    int shortest_index = -1;
    int shortest_burst = INT_MAX;

    for (int i = 0; i < n; i++) {
        if (!processes[i].completed && processes[i].arrival_time <= current_time) {
            if (processes[i].burst_time < shortest_burst) {
                shortest_burst = processes[i].burst_time;
                shortest_index = i;
            }
        }
    }

    return shortest_index;
}

// Function to perform SJF Non-Preemptive Scheduling
void sjf_non_preemptive(Process processes[], int n) {
    sort_by_arrival_and_burst(processes, n);

    int current_time = 0;
    int completed_processes = 0;

    float total_turnaround_time = 0;
    float total_waiting_time = 0;

    printf("Gantt Chart:\n");
    while (completed_processes < n) {
        int next_process = find_next_process(processes, n, current_time);

        if (next_process != -1) {
            Process *current = &processes[next_process];

            if (current_time < current->arrival_time) {
                current_time = current->arrival_time; // Advance time if CPU is idle
            }

            printf("| %s ", current->pid);
            int start_time = current_time;
            current_time += current->burst_time;
            int end_time = current_time;

            // Update process metrics
            current->completion_time = end_time;
            current->turnaround_time = current->completion_time - current->arrival_time;
            current->waiting_time = current->turnaround_time - current->burst_time;
            current->response_time = start_time - current->arrival_time;
            current->completed = true;

            total_turnaround_time += current->turnaround_time;
            total_waiting_time += current->waiting_time;

            completed_processes++;
        } else {
            current_time++; // No process ready, increment time
        }
    }
    printf("|\n");

    // Print results
    printf("\nPID\tArrival\tBurst\tCompletion\tTurnaround\tWaiting\tResponse\n");
    for (int i = 0; i < n; i++) {
        printf("%s\t%d\t%d\t%d\t\t%d\t\t%d\t%d\n",
               processes[i].pid,
               processes[i].arrival_time,
               processes[i].burst_time,
               processes[i].completion_time,
               processes[i].turnaround_time,
               processes[i].waiting_time,
               processes[i].response_time);
    }

    // Print averages
    printf("\nAverage Turnaround Time: %.2f\n", total_turnaround_time / n);
    printf("Average Waiting Time: %.2f\n", total_waiting_time / n);
}

int main() {
    int n;
    printf("Enter the number of processes: ");
    scanf("%d", &n);

    Process processes[n];
    for (int i = 0; i < n; i++) {
        printf("Enter Process ID, Arrival Time, and Burst Time for Process %d: ", i + 1);
        scanf("%s %d %d", processes[i].pid, &processes[i].arrival_time, &processes[i].burst_time);
        processes[i].completed = false;
    }

    sjf_non_preemptive(processes, n);

    return 0;
}

File: sjf(non).py
import matplotlib.pyplot as plt

class Process:
    def __init__(self, pid, arrival_time, burst_time):
        self.pid = pid
        self.arrival_time = arrival_time
        self.burst_time = burst_time
        self.completion_time = 0
        self.turnaround_time = 0
        self.waiting_time = 0

# Function to plot Gantt Chart
def plot_gantt_chart(gantt_chart):
    fig, gnt = plt.subplots()

    # Setting labels for x and y axis
    gnt.set_xlabel('Time')
    gnt.set_ylabel('Processes')

    # Set x-axis limits
    gnt.set_xlim(0, gantt_chart[-1][2])

    # Set y-axis ticks
    gnt.set_yticks([i + 1 for i in range(len(set([p[0] for p in gantt_chart])))])
    gnt.set_yticklabels([f"P{p[0]}" for p in gantt_chart])

    # Plotting the processes as bars in the Gantt chart
    for process, start_time, end_time in gantt_chart:
        gnt.broken_barh([(start_time, end_time - start_time)], (process, 0.9), facecolors='green')

    # Set chart title
    plt.title('Gantt Chart - SJF Non-preemptive Scheduling')
    plt.show()

# Function to simulate SJF Non-preemptive Scheduling
def sjf_non_preemptive_scheduling(processes):
    # Sort processes by Arrival Time initially
    processes.sort(key=lambda x: x.arrival_time)
    
    completed_processes = []
    ready_queue = []
    current_time = 0
    gantt_chart = []

    while processes or ready_queue:
        # Move processes that have arrived by current_time to ready_queue
        ready_queue += [p for p in processes if p.arrival_time <= current_time]
        processes = [p for p in processes if p.arrival_time > current_time]

        # Sort ready queue by burst time (shortest job first)
        ready_queue.sort(key=lambda x: x.burst_time)

        if ready_queue:
            # Select the process with the shortest burst time
            current_process = ready_queue.pop(0)
            start_time = current_time
            current_time += current_process.burst_time
            end_time = current_time

            # Update process details
            current_process.completion_time = end_time
            current_process.turnaround_time = current_process.completion_time - current_process.arrival_time
            current_process.waiting_time = current_process.turnaround_time - current_process.burst_time
            completed_processes.append(current_process)

            # Add to Gantt chart data
            gantt_chart.append((int(current_process.pid[1:]), start_time, end_time))  # Process ID without 'P'
        else:
            # If no process in ready queue, advance time to the next arrival
            current_time = processes[0].arrival_time

    # Print the results
    print("PID\tArrival\tBurst\tCompletion\tTurnaround\tWaiting")
    for process in completed_processes:
        print(f"{process.pid}\t{process.arrival_time}\t{process.burst_time}\t{process.completion_time}\t{process.turnaround_time}\t{process.waiting_time}")

    # Calculate average turnaround time and waiting time
    total_turnaround_time = sum([p.turnaround_time for p in completed_processes])
    total_waiting_time = sum([p.waiting_time for p in completed_processes])
    
    print(f"\nAverage Turnaround Time: {total_turnaround_time / len(completed_processes):.2f}")
    print(f"Average Waiting Time: {total_waiting_time / len(completed_processes):.2f}")

    # Plot Gantt chart
    plot_gantt_chart(gantt_chart)

# Input processes for SJF Non-preemptive
process_list = [
    Process("P1", 0, 8),
    Process("P2", 1, 4),
    Process("P3", 2, 9),
    Process("P4", 3, 5)
]

# Run SJF Non-preemptive simulation
sjf_non_preemptive_scheduling(process_list)
